Add P-EQDSK/pFile methods

LiamPattinson opened the pull request:
Adds `read` and `write` for P-EQDSK/pFiles, along with docs and tests.

`read` returns a named tuple containing `profiles`, `species`, and `units`. Both `profiles` and `species` are contained in Pandas DataFrames, while `units` is a dict storing the units for each column in `profiles`. All three are needed to write a new P-EQDSK file. Usage examples are in the README and docstrings.

I'd be grateful for any feedback on the implementation using Pandas. Some feedback on the function signatures would also be appreciated, because I'm not convinced that `read` returning a 3-tuple or `write` requiring 3 inputs is the cleanest solution.

It'd be useful for somebody to look over the docstrings I've written too, as there's a good chance they're inaccurate. @jfparisi If you have time, please could you skim over this and let me know if everything is correct?

Finally, any additional test data would be helpful, as while I have access to a vague specification and a single example file, I'm aware that the contents of P-EQDSK files can vary quite a lot, and it'd be good to know if these functions can handle any P-EQDSK file thrown at them.

## Known Issues:

- Whitespace isn't conserved when reading and re-writing. Hopefully this isn't a problem, but if codes that read P-EQDSK depend on a certain amount of whitespace at the start of each line and between each column, then the `write` function doesn't produce valid files.

LiamPattinson committed:
Add P-EQDSK read/write

LiamPattinson committed:
Add P-EQDSK tests

LiamPattinson committed:
Add P-EQDSK to README

LiamPattinson committed:
Add test for dimensionality of test data

LiamPattinson committed:
Minor README update

bpatel2107 commented:
Is this ready to merge? I would like to get https://github.com/pyro-kinetics/pyrokinetics/pull/179 merged if possible

LiamPattinson commented:
Had a chat with @ZedThree about this, and I'll need to make some further edits. Hopefully won't take too long to get this finished, but I imagine it'll take some significant further work on that other PR to get it over the finish line (tests, docs, cleanup, etc).

LiamPattinson committed:
Redesign P-EQDSK, removed Pandas dependency

Now returns built-in dicts/lists instead of Pandas DataFrames.
Uses TypedDict to assist with documentation/type hinting.

LiamPattinson commented:
@ZedThree I've rewritten the P-EQDSK read/write methods using `TypedDict` as you suggested, and removed the Pandas dependency. Is the closer to sort of design you were thinking of? Let me know if there's anything else I should change here.

ZedThree commented on freeqdsk/peqdsk.py:
```diff
@@ -0,0 +1,211 @@
+
+            yield "__species__", species
+
+        else:
```
Minor, but you don't need the `else` after `yield` (or `return`)

LiamPattinson replied:
I had to put a `continue` after it, as otherwise it would try to read a new profile right after finishing the species, but it does at least allow a single un-indent on the else block.

ZedThree replied:
Ah yes, sorry! I don't write generators myself very often

ZedThree commented on freeqdsk/peqdsk.py:
```diff
@@ -0,0 +1,211 @@
+
+        else:
+            # Get name and units
+            match = re.search(r"(.*)\((.*)\)", header[2])
```
Minor optimisation: pull the pattern out of the loop and `re.compile` it. If you want to be really fancy, you could even name the groups -- probably not quite worth it for such a simple regex though

LiamPattinson replied:
Done. For some reason it never occurs to me to pre-compile regex. Also never used named groups before, so that's a first.

LiamPattinson committed:
Minor edits to peqdsk _read_peqdsk_blocks
